96.md

```diff
@@ -0,0 +1,377 @@
+  }
+}
+```
+
+Note that if using a CDN, the `server` should issue a 302 redirect from
+the `nip96.routes.download` route to the CDN url instead
+of assigning the CDN directly as the `nip96.routes.download` route.
+This is because some NIP-96 integrations, such as NIP-60 zap gates,
+may require NIP-98 authentication for downloads,
+which the CDN won't be able to support.
+
+### List of Supporting File Storage Servers
+
+| Name | Domain |
+| ---- | ------ |
+| -    | -      |
+
+## Upload
+
+A file can be uploaded one at a time to `https://upload.domain/upload-path` (route from `https://server.example.domain/.well-known/nostr.json`) as `multipart/form-data` content type using `POST` method with the file object set to the `file` form data field.
+
+`Clients` must add a [NIP-98](98.md) `Authorization` header with the encoded `payload` tag set to the base64-encoded 256-bit SHA-256 hash of the file (not the hash of the whole request body).
+If using an html form, use an `Authorization` form data field instead.
+
+These following **optional** form data fields MAY be used by `servers` and SHOULD be sent by `clients`:
+- `expiration`: string of the UNIX timestamp in seconds. Empty string if file should be stored forever. The server isn't required to honor this;
+- `size`: string of the file byte size. This is just a value the server can use to reject early if the file size exceeds the server limits;
+- `alt`: strict description text for visibility-impaired users;
+- `caption`: loose description.
+
+Others custom form data fields may be used depending on specific `server` support.
+The `server` isn't required to store any metadata sent by `clients`.
+
+Note for `clients`: if using an HTML form, it is important for the `file` form field to be the **last** one, or be re-ordered right before sending or be appended as the last field of XHR2's FormData object.
+
+The `filename` embedded in the file may not be honored by the `server`, which could internally store just the SHA-256 hash value as the file name, ignoring extra metadata.
+The hash is enough to uniquely identify a file, that's why it will be used on the "download" and "delete" routes.
+
+The `server` MUST link the user's `pubkey` string (which is embedded in the decoded header value) as the owner of the file so to later allow him to delete the file.
+Note that if a file with the same hash of a previously received file (so the same file) is uploaded by another user, the server doesn't need to store the new file.
+It should just add the new user's `pubkey` to the list of the owners of the already stored file with said hash.
+
+The `server` SHOULD also store the `Authorization` header/field value (decoded or not) for accountability purpose as this proves that the user with the unique pubkey did ask for the upload.
+
+The `server` MUST reject with 413 Payload Too Large if file size exceeds limits.
+
+The `server` MUST reject with 401 Unauthorized status if some fiels are invalid.
```

Comment:
nit: This should probably be `400 Bad Request`

Reply:
Thanks. Fixed typo and changed to 422.

Reply:
FWIW I have never encountered a 422 in the wild. Pretty much everyone uses a 400 for Bad Requests.

Reply:
Ok changed 422 to 400